libs/frontend/feature-build/src/components/ComponentTree.vue

```diff
@@ -1,35 +1,167 @@
 <template>
-  <div class="component-tree" :class="{ 'component-tree-visible': show }">
-    <ComponentTreeItem :component="root" :level="0" :componentIndex="0" />
+  <div class="component-tree-wrap" :class="{ visible: show }">
+    <component :is="'style'" v-cloak>
+      {{ levelStyles }}
+    </component>
+    <div class="component-tree">
+      <ComponentTreeItem v-if="root" :component="root" :level="0" :componentIndex="0" />
+    </div>
   </div>
 </template>
 
 <script lang="ts" setup>
 import { computed } from 'vue'
 import { useBuild } from '@pubstudio/frontend/feature-build'
 import ComponentTreeItem from './ComponentTreeItem.vue'
+import { IComponent } from '@pubstudio/shared/type-site'
 
 const { editor, activePage } = useBuild()
 
 const root = computed(() => activePage.value?.root)
 
 const show = computed(() => editor.value?.showComponentTree)
+
+interface IterNode {
+  cmp: IComponent
+  depth: number
+}
+const makeNode = (component: IComponent, depth: number): IterNode => ({
+  cmp: component,
+  depth,
+})
+
+const levelStyles = computed(() => {
+  const styles: string[] = []
+  for (let i = 0; i <= treeDepth.value; i += 1) {
+    styles.push(`.cmp-title-${i} {padding-left: ${i * 12}px;}`)
+  }
+  return styles.join('\n')
+})
+
+const treeDepth = computed(() => {
+  const root = activePage.value?.root
+  const expandedCmp = editor.value?.componentTreeExpandedItems
+  if (!root || !expandedCmp) {
+    return 0
+  }
+  let depth = 0
+  // Tree iteration
+  const stack = [makeNode(root, 0)]
+  while (stack.length > 0) {
+    const node = stack.pop()
+    if (node && expandedCmp[node.cmp.id]) {
+      if (node.depth > depth) {
+        depth = node.depth
+      }
+      if (node.cmp.children) {
+        stack.push(...node.cmp.children.map((c) => makeNode(c, depth + 1)))
+      }
+    }
+  }
+  return depth
+})
+
+const treeStyles = computed(() => {
```

Comment:
@abemscac I "fixed" the nesting problem by checking the deepest component level, and setting some styles based on that. It's kind of complicated and requires traversing most of the component tree to calculate it, but seems to work.

I also made some changes to (hopefully) improve performance. I moved all the CSS out of the `ComponentTreeItem`, so it only needs to be processed once. Since we know the max tree depth, it's also possible to dynamically create styles for all the title paddings at each level.

While implementing this, I also re-enabled the eye/hide feature, although I'm not sure I like it.